contracts/evm/contracts/adapters/ClusterConnection.sol

```diff
@@ -0,0 +1,253 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.8.0;
+pragma abicoder v2;
+
+import "openzeppelin-contracts-upgradeable/contracts/proxy/utils/Initializable.sol";
+import "@xcall/utils/Types.sol";
+import "@xcall/contracts/xcall/interfaces/IConnection.sol";
+import "@iconfoundation/xcall-solidity-library/interfaces/ICallService.sol";
+
+contract ClusterConnection is Initializable, IConnection {
+    mapping(string => uint256) private messageFees;
+    mapping(string => uint256) private responseFees;
+    mapping(string => mapping(uint256 => bool)) receipts;
+    mapping(address => bool) public isValidator;
+
+    address private xCall;
+    address private adminAddress;
+    uint256 public connSn;
+    address[] private validators;
+    uint8 private reqValidatorCnt;
+
+    event Message(string targetNetwork, uint256 sn, bytes _msg);
+    event ValidatorAdded(address _validator);
+    event ValidatorRemoved(address _validator);
+
+    modifier onlyAdmin() {
+        require(msg.sender == this.admin(), "OnlyRelayer");
+        _;
+    }
+
+    function initialize(address _relayer, address _xCall) public initializer {
+        xCall = _xCall;
+        validators.push(_relayer);
+        adminAddress = _relayer;
+        emit ValidatorAdded(_relayer);
+    }
+
+    function listValidators() external view returns (address[] memory) {
+        return validators;
+    }
+
+    function addValidator(address _validator) external onlyAdmin {
+        require(!isValidator[_validator], "Address is already an signer");
+        validators.push(_validator);
+        isValidator[_validator] = true;
+        emit ValidatorAdded(_validator);
+    }
+
+    function removeValidator(address _validator) external onlyAdmin {
+        require(_validator!=this.admin(), "cannot remove admin");
```

Comment:
check for validators length should always be greater than threshold, and also, revert if validators list doesn't contain validator to be removed, revert will save gas fee

Reply:
updated